dbms/src/Interpreters/Join.cpp

```diff
@@ -582,7 +585,15 @@ class AddedColumns
                  const Block & block_with_columns_to_add,
                  const Block & block,
                  const Block & saved_block_sample,
-                 const ColumnsWithTypeAndName & extras)
+                 const ColumnsWithTypeAndName & extras,
+                 const Join & join_,
+                 const ColumnRawPtrs & key_columns_,
+                 const Sizes & key_sizes_)
+        : join(join_)
```

Comment:
Why do we need all new added fields? It doesn't seem we use it in this class.

Reply:
It's to avoid passing them through switchJoinRightColumns() -> joinRightColumns() call stack as function arguments. In fact it could have small performance penalty to pack them as a struct fields, cause we do not use all avaliable registers for fastcall. But it was unreadable code with messy function signatures in that place. We could revert it back when we don't need to read this code anymore :)

Reply:
Ah, I see. just haven't noticed they are public.